kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/AddFile.java

```diff
@@ -93,4 +103,208 @@ public static Row convertDataFileStatus(
     // any fields not present in the valueMap are considered null
     return new GenericRow(FULL_SCHEMA, valueMap);
   }
+
+  /**
+   * Utility to generate an {@link AddFile} action from an 'AddFile' {@link Row}.
+   *
+   * @throws NullPointerException if row is null
+   */
+  public static AddFile fromRow(Row row) {
+    Objects.requireNonNull(row, "Cannot generate an AddFile action from a null row");
+
+    checkArgument(
+        row.getSchema().equals(FULL_SCHEMA),
+        "Expected schema: %s, found: %s",
+        FULL_SCHEMA,
+        row.getSchema());
+
+    return new AddFile(
+        InternalUtils.requireNonNull(row, COL_NAME_TO_ORDINAL.get("path"), "path")
+            .getString(COL_NAME_TO_ORDINAL.get("path")),
+        InternalUtils.requireNonNull(
+                row, COL_NAME_TO_ORDINAL.get("partitionValues"), "partitionValues")
+            .getMap(COL_NAME_TO_ORDINAL.get("partitionValues")),
+        InternalUtils.requireNonNull(row, COL_NAME_TO_ORDINAL.get("size"), "size")
+            .getLong(COL_NAME_TO_ORDINAL.get("size")),
+        InternalUtils.requireNonNull(
+                row, COL_NAME_TO_ORDINAL.get("modificationTime"), "modificationTime")
+            .getLong(COL_NAME_TO_ORDINAL.get("modificationTime")),
+        InternalUtils.requireNonNull(row, COL_NAME_TO_ORDINAL.get("dataChange"), "dataChange")
+            .getBoolean(COL_NAME_TO_ORDINAL.get("dataChange")),
+        Optional.ofNullable(
+            row.isNullAt(COL_NAME_TO_ORDINAL.get("deletionVector"))
+                ? null
+                : DeletionVectorDescriptor.fromRow(
+                    row.getStruct(COL_NAME_TO_ORDINAL.get("deletionVector")))),
+        Optional.ofNullable(
+            row.isNullAt(COL_NAME_TO_ORDINAL.get("tags"))
+                ? null
+                : row.getMap(COL_NAME_TO_ORDINAL.get("tags"))),
+        Optional.ofNullable(
+            row.isNullAt(COL_NAME_TO_ORDINAL.get("baseRowId"))
+                ? null
+                : row.getLong(COL_NAME_TO_ORDINAL.get("baseRowId"))),
+        Optional.ofNullable(
+            row.isNullAt(COL_NAME_TO_ORDINAL.get("defaultRowCommitVersion"))
+                ? null
+                : row.getLong(COL_NAME_TO_ORDINAL.get("defaultRowCommitVersion"))),
+        Optional.ofNullable(
+            row.isNullAt(COL_NAME_TO_ORDINAL.get("stats"))
+                ? null
+                : DataFileStatistics.deserializeFromJson(
+                        row.getString(COL_NAME_TO_ORDINAL.get("stats")))
+                    .orElse(null)));
+  }
+
+  private final String path;
```

Comment:
@vkorukanti -- were we trying to avoid materializing all of the AddFile fields in memory into a POJO?

the addfile is materialized into a pojo inside of `RowTracking.java` below like so:

```java
                  AddFile addFile = AddFile.fromRow(row.getStruct(ADD_FILE_ORDINAL));

                  // Assign base row ID if missing
                  if (!addFile.getBaseRowId().isPresent()) {
                    final long numRecords = getNumRecordsOrThrow(addFile);
                    addFile = addFile.withNewBaseRowId(currRowIdHighWatermark.get() + 1L);
                    currRowIdHighWatermark.addAndGet(numRecords);
                  }

                  // Assign default row commit version if missing
                  if (!addFile.getDefaultRowCommitVersion().isPresent()) {
                    addFile = addFile.withNewDefaultRowCommitVersion(commitVersion);
                  }

                  // Return a new AddFile row with assigned baseRowId/defaultRowCommitVersion
                  return SingleAction.createAddFileSingleAction(addFile.toRow());
```

Reply:
I think there is a better way here, to implement AddFile without materializing all of the values, and just pointing to the underlying row

Reply:
I think we can solve this problem like so:

## Step 1: Create a DelegateRow class

```java
public abstract class DelegateRow implements Row {
    private final Row delegate;

    public DelegateRow(Row delegate) {
        this.delegate = delegate;
    }

    @Override
    public StructType getSchema() {
        return baseRow.getSchema();
    }

    // implement all Row interfaces using the delegate
}
```

## Step 2: Implement AddFile class without materializing everything

```java
class AddFile {
  private final Row row;

  public AddFile(Row row) {
    this.row = row;

    this.parsedPartitionValues = ...
  }

  public String getPath() {
    return row.getString(COL_NAME_TO_ORDINAL.get("path"));
  }

  // ... implement getters by referencing the row ...

  public AddFile withNewBaseRowId(long baseRowId) {
    Row updatedRow = new DelegateRow(row) {
      @Override
      public long getLong(int ordinal) {
        if (ordinal == COL_NAME_TO_ORDINAL.get("baseRowId")) {
          return baseRowId;
        }
        return super.getLong(ordinal);
      }

      @Override
      public boolean isNullAt(int ordinal) {
        if (ordinal == COL_NAME_TO_ORDINAL.get("baseRowId")) {
          return false; // baseRowId is now defined
        }
        return super.isNullAt(ordinal);
      }
    };
    return new AddFile(updatedRow);
  }
}
```

Reply:
Another way to implement DelegateRow is to pass in a map of overrides ... I'd have to think more of the tradeoffs between the two designs

Reply:
Yes we should avoid materializing all of the AddFile fields in memory into a POJO

Reply:
Is the issue that we now need to modify an add file? Instead of just previously accessing the values?

Reply:
Is the issue that we now need to modify an add file? Instead of just previously accessing the values/writing them as is?

I think this is important to think about from generic rust/java standpoint as well. Like if we have some generic chunk of engine data, and we need to change/update/add to it, how do we do so?

Reply:
Can we unblock this change and go forward with the current approach while we're figuring things out?

I agree it's not ideal to materialize AddFiles, but we only do it when row tracking is used so this PR is still a strict improvement over the current situation since tables with row tracking are not readable by kernel today.

Avoiding materializing actions to mutate them is then an optimization. It may also be useful for other features in the future

Reply:
Hi @johanl-db and @qiyuandong-db -- I think we need to pause work on this PR and take a step back to really plan this out. I think that the Kernel folks need to decide how we want (or further, _if_ we want) Rows to be created and updated.

For example, another way to accomplish this would be to create an expression that trasforms the row to create a new one with whatever values we want.